pylot/planning/rrt_star/rrt_star_planning_operator.py

```diff
@@ -0,0 +1,155 @@
+from pylot.simulation.utils import Transform, Location, Rotation
+from pylot.simulation.carla_utils import get_map
+from pylot.planning.rrt_star.rrt_star import apply_rrt_star
+from pylot.planning.rrt_star.utils import start_target_to_space
+from pylot.planning.utils import get_waypoint_vector_and_angle
+from pylot.utils import is_within_distance_ahead
+
+
+DEFAULT_OBSTACLE_LENGTH = 3  # 3 meters from front to back
+DEFAULT_OBSTACLE_WIDTH = 2  # 2 meters from side to side
+DEFAULT_TARGET_LENGTH = 1  # 1.5 meters from front to back
+DEFAULT_TARGET_WIDTH = 1  # 1 meters from side to side
+DEFAULT_DISTANCE_THRESHOLD = 10  # 10 meters ahead of ego
+DEFAULT_NUM_WAYPOINTS = 50  # 50 waypoints to plan for
+
+
+class RRTStarPlanningOperator(Op):
+    """ RRTStar Planning operator for Carla 0.9.x.
+
+    IMPORTANT: Do not use with older Carla versions.
+    The operator either receives all the waypoints from the scenario runner
+    agent (on the global trajectory stream), or computes waypoints using the
+    HD Map.
+    """
+    def __init__(self,
+                 name,
+                 flags,
+                 goal_location=None,
+                 log_file_name=None,
+                 csv_file_name=None):
+        super(RRTStarPlanningOperator, self).__init__(name)
+        self._log_file_name = log_file_name
+        self._logger = setup_logging(self.name, log_file_name)
+        self._csv_logger = setup_csv_logging(self.name + '-csv', csv_file_name)
+        self._flags = flags
+
+        self._wp_index = 9
+        self._vehicle_transform = None
+        self._waypoints = None
+        self._map = HDMap(get_map(self._flags.carla_host,
+                                  self._flags.carla_port,
+                                  self._flags.carla_timeout),
+                          log_file_name)
+        self._goal_location = carla.Location(*goal_location)
+
+        self._can_bus_msgs = deque()
+        self._prediction_msgs = deque()
+        self.msg_buffers = [self._can_bus_msgs, self._prediction_msgs]
+
+        self._lock = threading.Lock()
+
+    @staticmethod
+    def setup_streams(input_streams):
+        input_streams.filter(pylot.utils.is_can_bus_stream).add_callback(
+            RRTStarPlanningOperator.on_can_bus_update)
+        input_streams.filter(pylot.utils.is_prediction_stream).add_callback(
+            RRTStarPlanningOperator.on_prediction_update)
+        input_streams.add_completion_callback(
+            RRTStarPlanningOperator.on_notification)
+        return [pylot.utils.create_waypoints_stream()]
+
+    def on_can_bus_update(self, msg):
+        with self._lock:
+            self._can_bus_msgs.append(msg)
+
+    def on_prediction_update(self, msg):
+        with self._lock:
+            self._prediction_msgs.append(msg)
+
+    def on_notification(self, msg):
+        # get ego info
+        can_bus_msg = self._can_bus_msgs.popleft()
+        self._vehicle_transform = can_bus_msg.data.transform
```

Comment:
It looks like vehicle_transform shouldn't be part of the operator state because it is only used in this method and in update waypoints, which should probably be merged in on_notification.

Reply:
Sounds good, I'll merge the update waypoints into on_notification and remove vehicle_transform as a class attribute.